src/main/java/net/neoforged/neoforge/network/event/RegisterPayloadHandlerEvent.java

```diff
@@ -5,31 +5,70 @@
+ * <li>Payloads must be registered before they can be sent over the network.</li>
+ * <li>Payloads may only be sent over the {@link ConnectionProtocol} they were registered for.</li>
+ * <li>By default, {@link IPayloadRegistrar} will wrap handlers so they execute on the main thread instead of the network thread.</li>
+ * <li>Payloads are always serialized, including for {@link Connection#isMemoryConnection() memory connections} (singleplayer and LAN hosts).</li>
+ * <ul>
  */
 public class RegisterPayloadHandlerEvent extends Event implements IModBusEvent {
-    private final Function<String, IPayloadRegistrar> registrarFactory;
+    private final NetworkRegistry registry;
 
     @ApiStatus.Internal
-    public RegisterPayloadHandlerEvent(Function<String, IPayloadRegistrar> registrarFactory) {
-        this.registrarFactory = registrarFactory;
+    public RegisterPayloadHandlerEvent(NetworkRegistry registry) {
+        this.registry = registry;
+    }
+
+    /**
+     * Creates a new {@link IPayloadRegistar}, a utility for registering payloads using a builder-style format.
+     * 
+     * @param version The network version
+     */
+    public IPayloadRegistrar registrar(String version) {
+        return new PayloadRegistrar(version, this);
     }
 
     /**
-     * {@return A {@link IPayloadRegistrar} for the given namespace, creating one if it doesn't exist.}
+     * Registers a new payload.
+     * 
+     * @param <T>       The class of the payload.
+     * @param <B>       The class of the ByteBuf. Only {@link ConnectionProtocol#PLAY play} payloads may use {@link RegistryFriendlyByteBuf}.
+     * @param type      The type of the payload.
+     * @param codec     The codec for the payload.
+     * @param handler   The handler for the payload. This handler should expect to receive the payload on all declared protocols and flows. It will be executed on the network thread.
+     * @param protocols The protocols this payload supports being sent over. Only {@link ConnectionProtocol#CONFIGURATION configuration} and {@link ConnectionProtocol#PLAY play} are supported.
+     * @param flow      The flow of this payload. Specify {@link Optional#empty()} to support sending in both directions.
+     * @param version   The version of the payload. Increase the payload version if the codec logic or handler logic changes. Neo-Neo connections with mismatched versions are denied.
+     * @param optional  If the payload is optional. Any connection with missing non-optional payloads is denied.
+     * @throws UnsupportedOperationException if the registration fails for any reason.
      */
-    public IPayloadRegistrar registrar(String namespace) {
-        return registrarFactory.apply(namespace);
+    public <T extends CustomPacketPayload, B extends FriendlyByteBuf> void register(CustomPacketPayload.Type<T> type, StreamCodec<? super B, T> codec, IPayloadHandler<T> handler,
```

Comment:
I suggest we remove it and let all registration happen through the registrar. Then we can also remove the `NetworkRegistry` field and make `NetworkRegistry#register` static and package-private.